### PR TITLE
[RelayerSessions] feat: add the interfaces for the RelayerSessions and SessionTree

### PR DESCRIPTION
**Pull request by red-0ne**
### Human Summary

Add the interfaces for the `RelayerSessions` (SessionManager) and `SessionTree` that wraps an SMST with its corresponding session.

### Summary generated by Reviewpad on 07 Nov 23 09:43 UTC
This pull request includes the following changes:

- Added the module "github.com/cosmos/cosmos-proto" with version v1.0.0-beta.2
- Added the module "github.com/pokt-network/smt" with version v0.7.1
- Added the module "github.com/dgraph-io/badger/v4" with version v4.2.0
- Removed the module "github.com/cosmos/cosmos-proto" with version v1.0.0-beta.2
- Removed the module "github.com/dgraph-io/badger/v4" with version v4.2.0
- Modified the file "pkg/relayer/session/interface.go" by adding new code

Please review these changes carefully.

## Type of change

Select one or more:

- [x] New feature, functionality or library
- [ ] Bug fix
- [ ] Code health or cleanup
- [ ] Documentation
- [ ] Other (specify)

## Testing

- [ ] **Run all unit tests**: `make go_develop_and_test`
- [ ] **Verify Localnet manually**: See the instructions [here](TODO: add link to instructions)

## Sanity Checklist

- [ ] I have tested my changes using the available tooling
- [x] I have performed a self-review of my own code
- [x] I have commented my code, updated documentation and left TODOs throughout the codebase

### CODE REVIEW

File: pkg/relayer/session/interface.go
Comment (Olshansk) [inline pkg/relayer/session/interface.go:12]:
```suggestion
// and retrieval of SMSTs for a given session. It can also be thought of as a SessionManager but 
// dedicated to Relayer behaviour.
```

File: pkg/relayer/session/interface.go
Comment (Olshansk) [inline pkg/relayer/session/interface.go:16]:
```suggestion

// TODO: Add `ExpiringSessions` to notify of sessions whose proof submission window is closing.

```

File: pkg/relayer/session/interface.go
Comment (Olshansk) [inline pkg/relayer/session/interface.go:14]:
```suggestion
	// ClosingSessions returns an observable that notifies of sessions ready to be claimed.
```

File: pkg/relayer/session/interface.go
Comment (Olshansk) [inline pkg/relayer/session/interface.go:19]:
```suggestion
	// it creates a new SMST for it before returning it.
	// An error is returned if the corresponding KVStore for SMST fails to be created.
```

File: pkg/relayer/session/interface.go
Comment (Olshansk) [inline pkg/relayer/session/interface.go:24]: Great design!

File: pkg/relayer/session/interface.go
Comment (Olshansk) [inline pkg/relayer/session/interface.go:30]:
```suggestion
	Update(key, value []byte, weight uint64) error
```

File: pkg/relayer/session/interface.go
Comment (Olshansk) [inline pkg/relayer/session/interface.go:38]:
```suggestion
	// Close gets the root hash of the SMST needed for submitting the claim; then commits the entire tree to disk and stops the KVStore.
```

File: pkg/relayer/session/interface.go
Comment (Olshansk) [inline pkg/relayer/session/interface.go:39]:
```suggestion
	// It should be called before submitting the claim on-chain. This function frees up the in-memory resources used by the SMST
```

File: pkg/relayer/session/interface.go
Comment (Olshansk) [inline pkg/relayer/session/interface.go:41]: What do you think of renaming this to `Flush`?

File: pkg/relayer/session/interface.go
Comment (Olshansk) [inline pkg/relayer/session/interface.go:36]: If we rename it to `Flush`, then it's a bit more reasonable and we can remove this comment, right?

File: pkg/relayer/session/interface.go
Comment (Olshansk) [inline pkg/relayer/session/interface.go:37]: Make sure the comments below aren't too long after my updates.

**Review by Olshansk — CHANGES_REQUESTED** [review]

File: pkg/relayer/session/interface.go
Comment (red-0ne) [reply pkg/relayer/session/interface.go:17]: Adding it to the interface. Was planning to handle it a bit differently but I think this is a better design.

Please confirm (form its comment) that this is what you intended it to be.

File: pkg/relayer/session/interface.go
Comment (Olshansk) [inline pkg/relayer/session/interface.go:41]:
```suggestion
	// Flush gets the root hash of the SMST needed for submitting the claim; then commits the entire tree to disk
```

**Review by Olshansk — APPROVED** [review]
@red-0ne Few things:

1. 1 minor NIT
2. In the future, feel free to resolve trivial comments if you've addressed them
3. Let's wait for the base branch to merge in, so we can squash & merge this one into main

Otherwise, LGTM!

**Review by Olshansk — CHANGES_REQUESTED** [review]
Will re-approve after merging with `main` and last NITs are resolved.

File: pkg/relayer/session/interface.go
Comment (bryanchriswhite) [inline pkg/relayer/session/interface.go:14]: :ok_hand: nice improvement from poktroll-alpha!

File: pkg/relayer/session/interface.go
Comment (bryanchriswhite) [inline pkg/relayer/session/interface.go:13]: Is `SessionManager` still a thing? The last sentence raises more questions than it does answers for me. Could we revise it to avoid referencing unrelated or nonexistent objects, or just drop it altogther?

Wdyt about also including expiring sessions in the comment if we're going to mention closing sessions, e.g.:

```suggestion
// RelayerSessions is an interface for managing the relayer's sessions and Sparse
// Merkle Sum Trees (SMSTs). It provides notifications about closing and expiring
// sessions that are ready to be claimed and proven, respectively. It also handles
// the creation and retrieval of SMSTs for a given session.
```

File: pkg/relayer/session/interface.go
Comment (bryanchriswhite) [inline pkg/relayer/session/interface.go:16]: Wdyt about `ClaimableSessions()` and `ProvableSessions()`?

File: pkg/relayer/session/interface.go
Comment (bryanchriswhite) [inline pkg/relayer/session/interface.go:1]: I think we need to move this file up one directory so that is is in the `/pkg/relayer` package. This will mitigate future import cycles caused by the interface and its implementation sharing a package.

Additionally, this has the effect of consolidating all of the interfaces which are dependencies of relayer into a single place.

File: pkg/relayer/session/interface.go
Comment (red-0ne) [reply pkg/relayer/session/interface.go:13]: Expiring sessions is removed on #105

File: pkg/relayer/session/interface.go
Comment (red-0ne) [reply pkg/relayer/session/interface.go:16]: Will port this change to #105

**Comment by red-0ne** [comment]
Closed in favor of #105

File: pkg/relayer/session/interface.go
Comment (Olshansk) [reply pkg/relayer/session/interface.go:1]: ➕ 1

**Pull request closed without merging**
